paddle/fluid/operators/conv_cudnn_helper.h

```diff
@@ -0,0 +1,271 @@
+
+namespace paddle {
+namespace operators {
+
+using framework::AlgorithmsCache;
+
+struct ConvArgs {
+  cudnnHandle_t handle;
+  platform::TensorDescriptor idesc, odesc;
+  platform::FilterDescriptor wdesc;
+  platform::ConvolutionDescriptor cdesc;
+  const framework::Tensor *x, *w, *o;
+
+  // strides
+  std::vector<int> s;
+  // paddings
+  std::vector<int> p;
+  // dilations
+  std::vector<int> d;
+
+  ConvArgs(const framework::Tensor* x, const framework::Tensor* w,
+           const framework::Tensor* o, const std::vector<int> s,
+           const std::vector<int> p, const std::vector<int> d)
+      : x(x), w(w), o(o), s(s), p(p), d(d) {}
+};
+
+template <typename perf_t>
+struct SearchAlgorithm {};
+
+template <>
+struct SearchAlgorithm<cudnnConvolutionFwdAlgoPerf_t> {
+  using perf_t = cudnnConvolutionFwdAlgoPerf_t;
+  using algo_t = cudnnConvolutionFwdAlgo_t;
+
+  template <typename T>
+  static algo_t find(const ConvArgs& args, bool exhaustive_search,
+                     bool deterministic, int algo_cache_id,
+                     const framework::ExecutionContext& ctx) {
+    auto dtype = platform::CudnnDataType<T>::type;
+    bool exhaustive = (exhaustive_search) & (dtype != CUDNN_DATA_HALF);
+
+    size_t workspace_size_limit = FLAGS_conv_workspace_size_limit * 1024 * 1024;
+
+    algo_t algo;
+    if (!exhaustive) {
+      CUDNN_ENFORCE(platform::dynload::cudnnGetConvolutionForwardAlgorithm(
+          args.handle, args.idesc.desc(), args.wdesc.desc(), args.cdesc.desc(),
+          args.odesc.desc(), CUDNN_CONVOLUTION_FWD_SPECIFY_WORKSPACE_LIMIT,
+          workspace_size_limit, &algo));
+      VLOG(3) << "choose algo " << algo;
+    } else {
+      AlgorithmsCache<algo_t>& algo_cache =
+          ctx.GetKernelConfig<AlgorithmsCache<algo_t>>(algo_cache_id);
+      auto& dev_ctx =
+          ctx.template device_context<platform::CUDADeviceContext>();
+      auto workspace_handle = dev_ctx.cudnn_workspace_handle();
+
+      auto x_dims = framework::vectorize(args.x->dims());
+      auto w_dims = framework::vectorize(args.w->dims());
+
+      algo = algo_cache.GetAlgorithm(
+          x_dims, w_dims, args.s, args.p, args.d, 0, [&]() {
+            int returned_algo_count;
+            std::array<perf_t, kNUM_CUDNN_FWD_ALGS> perf_stat;
+
+            auto cudnn_find_func = [&](void* cudnn_workspace_ptr) {
+              CUDNN_ENFORCE(
+                  platform::dynload::cudnnFindConvolutionForwardAlgorithmEx(
+                      args.handle, args.idesc.desc(), args.x->data<T>(),
+                      args.wdesc.desc(), args.w->data<T>(), args.cdesc.desc(),
+                      args.odesc.desc(), const_cast<T*>(args.o->data<T>()),
+                      kNUM_CUDNN_FWD_ALGS, &returned_algo_count,
+                      perf_stat.data(), cudnn_workspace_ptr,
+                      workspace_size_limit));
+            };
+            workspace_handle.RunFunc(cudnn_find_func, workspace_size_limit);
```

Comment:
Please call `RunFuncSync`, so that workspace call be released immediately after  `cudnnFindConvolutionForwardAlgorithmEx ` is called.